## Fetching Data from an API and a Local Spreadsheet
The datasets that I used were from [Sparrow](https://sparrow-data.org/labs/wiscar/api/v1/)  and from a spreadsheet on my local computer. To fetch data from the Sparrow API, in Python, I used the [Requests library](https://requests.readthedocs.io/en/master/). The data is a JSON object.

First thing we need to do is import all of the libraries used in this script. I'm using requests to fetch the JSON object, json which is included with Python and doesn't require a external download, and pandas. Pandas is used to handle large datasets in easy to manipulate dataframes. [Pandas](https://pandas.pydata.org/) is by far the most used library in this collection of notebooks.

In [1]:
import requests
import json
import pandas as pd

In [2]:
r = requests.get('https://sparrow-data.org/labs/wiscar/api/v1/sample?all=1')
online_metadata = r.json()
df1 = pd.DataFrame(online_metadata)
df1

id  igsn          name    material  \
0       18  None           M2C  Lava Flows   
1       19  None       90T151A       Baslt   
2       20  None       90T050B       Baslt   
3       21  None      84C207AB        None   
4       22  None   LDMEB-13-21      Dacite   
...    ...   ...           ...         ...   
1814  4846  None  TN182-03-004      Dacite   
1815  4847  None  TN182-07-004      Dacite   
1816  4848  None  TN182-08-014      Basalt   
1817  4849  None  TN182-09-001      Basalt   
1818  4850  None      UT-08-03        None   

                                               geometry location_name  \
0     {'type': 'Point', 'coordinates': [-149.66, -17...          None   
1     {'type': 'Point', 'coordinates': [-156.2311, 2...          None   
2     {'type': 'Point', 'coordinates': [-156.2311, 2...          None   
3                                                  None          None   
4     {'type': 'Point', 'coordinates': [-70.5921, -3...          None   
...                                                 ...           ...   
1814  {'type': 'Point', 'coordinates': [174.8567, 52...          None   
1815  {'type': 'Point', 'coordinates': [175.2449, 52...          None   
1816  {'type': 'Point', 'coordinates': [175.2772, 52...          None   
1817  {'type': 'Point', 'coordinates': [175.1453, 52...          None   
1818   {'type': 'Point', 'coordinates': [-110.8, 39.2]}          None   

      location_precision location_name_autoset  project_id project_name  \
0                      0                  None         NaN         None   
1                      0                  None         NaN         None   
2                      0                  None         NaN         None   
3                      0                  None         NaN         None   
4                      0                  None         NaN         None   
...                  ...                   ...         ...          ...   
1814                   0                  None         NaN         None   
1815                   0                  None         NaN         None   
1816                   0                  None         NaN         None   
1817                   0                  None         NaN         None   
1818                   0                  None         NaN         None   

      is_public  
0          True  
1          True  
2          True  
3          True  
4          True  
...         ...  
1814       True  
1815       True  
1816       True  
1817       True  
1818       True  

[1819 rows x 11 columns]

So what just happened? A number of things:
   * We used the `request.get()` function to fetch the JSON object from the API
   * We then stored it as JSON in the variable `online_metadata`
   * Lastly we turned the JSON object into a pandas dataframe with the function `pd.DataFrame`

>__IMPORTANT__: Note that the columns of the DataFrame _df_ are the keywords from the JSON object! And the populated cells are the values associated with each keyword for the specific sub-object. For my data each individual JSON object is a geologic sample. Also, __Note__ there are 1819 rows to this dataframe, but we are only seeing 10 rows. This is because for very large dataframes pandas automatically shortens the dataframe you will see in the terminal. To see the entire dataframe simply add `pd.set_option('display.max_rows', None)` after you import the library. Also you can probably see that the __Geometry__ column has in itself a JSON object. This is because in the API the geometry is a nested object. We will work on splitting that column into a _latitude_ and _longitude_ later.

Now we will import a microsoft excel spreadsheet into a pandas dataframe.

In [3]:
df2 = pd.read_excel('/Users/casey_idz/Documents/Coding/Python/WiscAr-Sparrow-Metadata/WiscAr_metadata(thisone).xlsx') # Metadata from Spreadsheet page 1
df3 = pd.read_excel('/Users/casey_idz/Documents/Coding/Python/WiscAr-Sparrow-Metadata/WiscAr_metadata(thisone).xlsx', sheet_name='Additional')# metadata from spreadsheet page 2


To import an excel sheet into a pandas dataframe we will utilize the pandas [read_excel](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) function. There are many parameters that this function can take; however, the only two that we care about are the `'path to file'` and the `sheet_name`. You only need to mention the `sheet_name` if you have more than one. The defualt will import the first sheet automatically without specification.

In [4]:
df2

sparrow order     id sample_name                  technique Irradiation  \
0            1132.0  408.0     GA0205A  Ar/Ar Incremental Heating       UW100   
1             870.0  341.0     GD0103B  Ar/Ar Incremental Heating       UW100   
2            1131.0  407.0     GD0103B  Ar/Ar Incremental Heating       UW100   
3             965.0  983.0      KORATH  Ar/Ar Incremental Heating       UW100   
4            1146.0  984.0      GB1501  Ar/Ar Incremental Heating       UW100   
...             ...    ...         ...                        ...         ...   
1257          347.0  281.0     LDM249A  Ar/Ar Incremental Heating        UW99   
1258         1101.0  774.0      LDM500  Ar/Ar Incremental Heating        UW99   
1259         1102.0  285.0        LDM6  Ar/Ar Incremental Heating        UW99   
1260         1103.0  773.0        LDM6  Ar/Ar Incremental Heating        UW99   
1261          339.0  191.0       LDM60  Ar/Ar Incremental Heating        UW99   

           phase sample_name.1            lithology  latitude  longitude  ...  \
0     groundmass       GA0205A             andesite  16.08800  -61.76500  ...   
1     groundmass       GD0103B             andesite  16.08800  -61.76500  ...   
2     groundmass       GD0103B             andesite  16.08800  -61.76500  ...   
3      amphibole        KORATH  pyroclastic deposit   5.10000   35.86700  ...   
4     groundmass        GB1501               Basalt  11.61460   41.65316  ...   
...          ...           ...                  ...       ...        ...  ...   
1257  groundmass       LDM249A                  NaN -35.98662  -70.49422  ...   
1258  groundmass        LDM500                  NaN -36.17494  -70.39531  ...   
1259          WR          LDM6                  NaN       NaN        NaN  ...   
1260  groundmass          LDM6                  NaN       NaN        NaN  ...   
1261  groundmass         LDM60                  NaN       NaN        NaN  ...   

        year                               journal  \
0     2013.0  Geochemistry, Geophysics, Geosystems   
1     2013.0  Geochemistry, Geophysics, Geosystems   
2     2013.0  Geochemistry, Geophysics, Geosystems   
3     2014.0                    Quaternary Geology   
4        NaN                                   NaN   
...      ...                                   ...   
1257     NaN                                   NaN   
1258     NaN                                   NaN   
1259     NaN                                   NaN   
1260     NaN                                   NaN   
1261     NaN                                   NaN   

                                                  Title  \
0     Snapshot of the Matuyama_Brunhes reversal proc...   
1     Snapshot of the Matuyama_Brunhes reversal proc...   
2     Snapshot of the Matuyama_Brunhes reversal proc...   
3     An age for the Korath Range, Ethiopia and the ...   
4                                                   NaN   
...                                                 ...   
1257                                                NaN   
1258                                                NaN   
1259                                                NaN   
1260                                                NaN   
1261                                                NaN   

                                          doi_link Comments  \
0               https://doi.org/10.1002/ggge.20263      NaN   
1               https://doi.org/10.1002/ggge.20263      NaN   
2               https://doi.org/10.1002/ggge.20263      NaN   
3     https://doi.org/10.1016/j.quageo.2013.03.007      NaN   
4                                              NaN      NaN   
...                                            ...      ...   
1257                                           NaN      NaN   
1258                                           NaN      NaN   
1259                                           NaN      NaN   
1260                                           NaN      NaN   
1261     

In [5]:
df3

sparrow order  id sample_name  technique Irradiation phase  \
0               NaN NaN   SEG 03 32        NaN         NaN   NaN   
1               NaN NaN   SEG 03 44        NaN         NaN   NaN   
2               NaN NaN   SEG 03 66        NaN         NaN   NaN   
3               NaN NaN   SEG 03 03        NaN         NaN   NaN   
4               NaN NaN     SB87–56        NaN         NaN   NaN   
...             ...  ..         ...        ...         ...   ...   
1033            NaN NaN        S3C1        NaN         NaN   NaN   
1034            NaN NaN      AH3-LP        NaN         NaN   NaN   
1035            NaN NaN      AH43-1        NaN         NaN   NaN   
1036            NaN NaN        AH46        NaN         NaN   NaN   
1037            NaN NaN       AH38f        NaN         NaN   NaN   

     sample_name.1            lithology   latitude   longitude  ...  \
0        SEG 03 32        rhyolite dome  52.351166 -175.417000  ...   
1        SEG 03 44     dacitic ash flow  52.351166 -175.417000  ...   
2        SEG 03 66  andesitic lava flow  52.283300 -172.403300  ...   
3        SEG 03 03    dacitic lava flow  52.375000 -172.389166  ...   
4          SB87–56  rhyolitic lava flow  52.266833 -172.522166  ...   
...            ...                  ...        ...         ...  ...   
1033          S3C1    Olivine Leucitite  41.000000   13.000000  ...   
1034        AH3-LP    Olivine Leucitite  41.000000   13.000000  ...   
1035        AH43-1            Leucitite  41.000000   13.000000  ...   
1036          AH46            Leucitite  41.000000   13.000000  ...   
1037         AH38f    Olivine Leucitite  41.000000   13.000000  ...   

                                       Member GeoDeepDive ping?        author  \
0     Rhyolite flow in crater valley (7.5 ka)               NaN  Jicha et al.   
1                          Dacitic Ignimbrite               NaN  Jicha et al.   
2                           Lava Point Dacite               NaN  Jicha et al.   
3                       Finch Cove Rhyodacite               NaN  Jicha et al.   
4     Basaltic to Rhyolitic South Shore Lavas               NaN  Jicha et al.   
...                                       ...               ...           ...   
1033                                      NaN               NaN      G. Mario   
1034                                      NaN               NaN      G. Mario   
1035                                      NaN               NaN      G. Mario   
1036                                      NaN               NaN      G. Mario   
1037                                      NaN               NaN      G. Mario   

        year                              journal  \
0     2005.0  Earth and Planetary Science Letters   
1     2005.0  Earth and Planetary Science Letters   
2     2005.0  Earth and Planetary Science Letters   
3     2005.0  Earth and Planetary Science Letters   
4     2005.0  Earth and Planetary Science Letters   
...      ...                                  ...   
1033  2016.0                               Lithos   
1034  2016.0                               Lithos   
1035  2016.0                               Lithos   
1036  2016.0                               Lithos   
1037  2016.0                               Lithos   

                                                  Title  \
0     Contrasting timescales of crystallization and ...   
1     Contrasting timescales of crystallization and ...   
2     Contrasting timescales of crystallization and ...   
3     Contrasting timescales of crystallization and ...   
4     Contrasting timescales of crystallization and ...   
...                                                 ...   
1033  Paleozoic metasomatism at the origin of Medite...   
1034  Paleozoic metasomatism at the origin of Medite...   
1035  Paleozoic metasomatism at the origin of Medite...   
1036  Paleozoic metasomatism at the origin of Medite...   
1037  Paleozoic metasomatism at the origin of Medite...   

                 

#### Recap:
In this notebook we learned how to import data from an API using the requests library and then turned it into a pandas dataframe. We also learned how to import excel files and turn them into pandas dataframes.

In [6]:
%store df1
%store df2
%store df3

Stored 'df1' (DataFrame)
Stored 'df2' (DataFrame)
Stored 'df3' (DataFrame)
